OpenXmlPowerTools/DocumentBuilder.cs

```diff
@@ -1636,79 +1845,313 @@ private static void AddToIgnorable(XElement root, string v)
+          </w:r>
+        </w:p>
+      </w:docPartBody>
+    </w:docPart>
+  </w:docParts>
+</w:glossaryDocument>
+#endif
+
+        private static void CopyCustomXmlPartsForDataBoundContentControls(WordprocessingDocument sourceDocument, WordprocessingDocument newDocument, IEnumerable<XElement> newContent)
+        {
+            List<string> itemList = new List<string>();
+            foreach (string itemId in newContent
+                .Descendants(W.dataBinding)
+                .Select(e => (string)e.Attribute(W.storeItemID)))
+                if (!itemList.Contains(itemId))
+                    itemList.Add(itemId);
+            foreach (CustomXmlPart customXmlPart in sourceDocument.MainDocumentPart.CustomXmlParts)
+            {
+                OpenXmlPart propertyPart = customXmlPart
+                    .Parts
+                    .Select(p => p.OpenXmlPart)
+                    .Where(p => p.ContentType == "application/vnd.openxmlformats-officedocument.customXmlProperties+xml")
+                    .FirstOrDefault();
+                if (propertyPart != null)
+                {
+                    XDocument propertyPartDoc = propertyPart.GetXDocument();
+                    if (itemList.Contains(propertyPartDoc.Root.Attribute(DS.itemID).Value))
+                    {
+                        CustomXmlPart newPart = newDocument.MainDocumentPart.AddCustomXmlPart(customXmlPart.ContentType);
+                        newPart.GetXDocument().Add(customXmlPart.GetXDocument().Root);
+                        foreach (OpenXmlPart propPart in customXmlPart.Parts.Select(p => p.OpenXmlPart))
+                        {
+                            CustomXmlPropertiesPart newPropPart = newPart.AddNewPart<CustomXmlPropertiesPart>();
+                            newPropPart.GetXDocument().Add(propPart.GetXDocument().Root);
+                        }
+                    }
+                }
+            }
+        }
+
+        private static Dictionary<XName, XName[]> RelationshipMarkup = null;
+
+        private static void UpdateContent(IEnumerable<XElement> newContent, XName elementToModify, string oldRid, string newRid)
+        {
+            foreach (var attributeName in RelationshipMarkup[elementToModify])
+            {
+                var elementsToUpdate = newContent
+                    .Descendants(elementToModify)
+                    .Where(e => (string)e.Attribute(attributeName) == oldRid);
+                foreach (var element in elementsToUpdate)
+                    element.Attribute(attributeName).Value = newRid;
+            }
+        }
+
+        private static void AddRelationships(OpenXmlPart oldPart, OpenXmlPart newPart, IEnumerable<XElement> newContent)
+        {
+            var relevantElements = newContent.DescendantsAndSelf()
+                .Where(d => RelationshipMarkup.ContainsKey(d.Name) &&
+                    d.Attributes().Any(a => RelationshipMarkup[d.Name].Contains(a.Name)))
+                .ToList();
```

Comment:
Unless you're going to do multiple checks that will evaluate the enumerable, there's no need to prematurely realize it. By having `ToList()` followed by where you just iterate over it causes the collection to be iterated over twice, wasting both some time and memory (as it has to be placed somewhere)

Reply:
Agreed.  I removed the ToList().